src/vm/methodtable.cpp

```diff
@@ -7059,24 +7117,62 @@ BOOL MethodTable::FindDefaultMethod(
                         );
                     }
 
-                    if (pBestCandidateMT != pCurMT)
-                    {                    
-                        if (pBestCandidateMT == NULL ||                         // first time
-                            pCurMT->CanCastToInterface(pBestCandidateMT))       // Prefer "more specific"" interface
+                    bool needToInsert = true;
+                    bool seenMoreSpecific = false;
+
+                    // We need to maintain the invariant that the candidates are always the most specific
+                    // in all path scaned so far. There might be multiple incompatible candidates 
+                    for (int i = 0; i < candidatesCount; ++i)
+                    {
+                        MethodTable *pCandidateMT = candidates[i].pMT;
+                        if (pCandidateMT == NULL)
+                            continue;
+
+                        if (pCandidateMT->HasSameTypeDefAs(pCurMT))
                         {
-                            // This is a better match
-                            pBestCandidateMT = pCurMT;
-                            pBestCandidateMD = pCurMD;
+                            // A dup - we are done
+                            needToInsert = false;
+                            break;
                         }
-                        else
+
+                        if (pCurMT->CanCastToInterface(pCandidateMT))
                         {
-                            if (!pBestCandidateMT->CanCastToInterface(pCurMT))
+                            // pCurMT is a more specific choice
+                            // If pCurMT is more specific than IFoo and IBar, only update first entry IFoo and null out IBar
+                            if (!seenMoreSpecific)
```

Comment:
The more specific is always the current pCurMT in this case (not array[4] or array[2]). So we effectively NULL-out all less specific entries and *assign* the first one to pCurMT.  There is no point keeping array[4] or array[2] at all - since pCurMT "survives".